10-habits/5_integration/internal/server/server.go

```diff
@@ -0,0 +1,74 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	_ "net/http/pprof"
+	"os"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/reflection"
+
+	"learngo-pockets/habits/api"
+	"learngo-pockets/habits/internal/habit"
+)
+
+// Server is the implementation of the gRPC server.
+type Server struct {
+	api.UnimplementedHabitsServer
+	db repository
+}
+
+type repository interface {
+	Add(ctx context.Context, habit habit.Habit) error
+	FindAll(ctx context.Context) ([]habit.Habit, error)
+}
+
+// New returns a Server that can Listen.
+func New(repo repository) *Server {
+	return &Server{
+		db: repo,
+	}
+}
+
+// Listen starts the listening to the port.
+func (s *Server) Listen(ctx context.Context, port int) error {
+	listener, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		return fmt.Errorf("unable to listen to tcp port %d: %w", port, err)
+	}
+
+	grpcServer := grpc.NewServer(grpc.UnaryInterceptor(timerInterceptor(os.Stdout)))
+	api.RegisterHabitsServer(grpcServer, s)
+	reflection.Register(grpcServer) // if env == dev
+	log.Printf("gRPC server started and listening to port %d", port)
+
+	errChan := make(chan error)
+	// Listen to the port. This will only return when something kills or stops the server.
+	go func() {
```

Comment:
The grpc server is shut down when the context is done by the parent function.

Reply:
When the context is done then it starts a signaling chain, which eventually starts shutting down the server. This is different from the server being completely shutdown and not processing any more requests.... but...

... You are partly right, because the code below is using `grpcServer.GracefulStop()` then that waits for the server to properly shut down before continuing.

But, initially I missed that this will leak a goroutine, which you could detect when you add an explicit wait for the goroutine (See "Know when things stop" in https://www.storj.io/blog/production-concurrency).

Reply:
Please add proper waiting to the goroutine, because https://github.com/alienorlatour/tiny-go-projects/pull/86/commits/f91cc6ca6739036ea3e7e18ec1837fc4a44c955e still does not fix the issue.

I don't want to repeat this; but all goroutines that you start should be waited on, otherwise it's really easy to make mistakes. I.e. you are not noticing the issue because of two reasons 1. you are not waiting for the goroutine to be completed before returning from this `Listen` and 2. you are missing tests for the server.

Reply:
Here is the full implementation: https://github.com/alienorlatour/tiny-go-projects/pull/96